### PR TITLE
Fix EmbeddedRocksDB with TTL writes corrupted data

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix a bug leads to EmbeddedRocksDB with TTL write corrupted SST files

When rocksdb database has TTL, a timestamp will be append to the end of value of each record. Nevertheless, `EmbeddedRocksDBBulkSink` generate SST files without the timestamp, so when ingesting to rocksdb database, the SST files are corrupted.

It's unfortunate that we didn't have CI test for `EmbeddedRocksDB` with TTL and cannot discover this bug before :(.

cc @vdimir 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit b0bdbda27e297f86616eccd6447166fa795ef4b4 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65816/b0bdbda27e297f86616eccd6447166fa795ef4b4/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/f3997640211ba3b29213191b9d6e194536233399/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/queries/0_stateless/02956_rocksdb_with_ttl.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
+-- Tags: no-ordinary-database, use-rocksdb
+
+CREATE TABLE dict_with_ttl (key UInt64, value String) ENGINE = EmbeddedRocksDB(2) PRIMARY KEY (key);
```
Comment (vdimir) [inline tests/queries/0_stateless/02956_rocksdb_with_ttl.sql:3]: Should TTL to be present if definition?

File: src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp [hunk 2]
```diff
@@ -169,14 +168,49 @@ std::pair<ColumnString::Ptr, ColumnString::Ptr> EmbeddedRocksDBBulkSink::seriali
...
+                memcpy(dst, &value, sizeof(value));
+            }
+            else
+            {
+                dst[0] = value & 0xff;
+                dst[1] = (value >> 8) & 0xff;
+                dst[2] = (value >> 16) & 0xff;
+                dst[3] = (value >> 24) & 0xff;
+            }
+        };
```
Comment (vdimir) [inline src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp:186]: Can we use `writeBinaryEndian` from `WriteHelpers`?

https://github.com/ClickHouse/ClickHouse/blob/ab0968740c6cd2ae4edd3601d7df03ff2970c192/src/IO/WriteHelpers.h#L1370-L1386

File: tests/queries/0_stateless/02956_rocksdb_with_ttl.sql (hunk 1 above)
Comment (canhld94) [reply tests/queries/0_stateless/02956_rocksdb_with_ttl.sql:3]: For `EmbeddedRocksDB` ttl is the first argument of the engine definition. I don't know why it was implemented this way in the first place, but it was.

**Comment by canhld94** [comment]
I think Bugfix validation is broken, it cannot detect new tests - https://github.com/ClickHouse/ClickHouse/issues/65946
Stateless:

```
00992_system_parts_race_condition_zookeeper_long
02112_delayed_clickhouse_local_with_queries_file
01007_r1r2_w_r2r1_deadlock
01560_ttl_remove_empty_parts
02319_lightweight_delete_on_merge_tree
01890_jit_aggregation_function_sum_long
02417_opentelemetry_insert_on_distributed_table
02521_incorrect_dealy_for_insert_bug_44902
01880_materialized_view_to_table_type_check
```

Fails because of `LOGICAL_ERROR` in `VerticalMerge`

```
2024-07-01 09:03:19 Thread 1025 "TCPHandler" received signal SIGABRT, Aborted.
2024-07-01 09:03:19 [Switching to Thread 0x7f3cbe379640 (LWP 5893)]
2024-07-01 09:03:19 0x00007f3fde5549fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
2024-07-01 09:03:19 #0  0x00007f3fde5549fc in pthread_kill () from /lib/x86_64-linux-gnu/libc.so.6
2024-07-01 09:03:19 No symbol table info available.
2024-07-01 09:03:19 #1  0x00007f3fde500476 in raise () from /lib/x86_64-linux-gnu/libc.so.6
2024-07-01 09:03:19 No symbol table info available.
2024-07-01 09:03:19 #2  0x00007f3fde4e67f3 in abort () from /lib/x86_64-linux-gnu/libc.so.6
2024-07-01 09:03:19 No symbol table info available.
2024-07-01 09:03:19 #3  0x0000562e70bbc4ff in __interceptor_abort ()
2024-07-01 09:03:19 No symbol table info available.
2024-07-01 09:03:19 #4  0x0000562e78fcfe05 in DB::abortOnFailedAssertion (description=...) at ./build_docker/./src/Common/Exception.cpp:44
2024-07-01 09:03:19 No locals.
2024-07-01 09:03:19 #5  0x0000562e78fd0b04 in DB::handle_error_code (msg=..., code=49, trace=..., remote=<optimized out>) at ./build_docker/./src/Common/Exception.cpp:61
2024-07-01 09:03:19 No locals.
2024-07-01 09:03:19 #6  DB::Exception::Exception (this=0x7250039f1880, msg_masked=..., code=49, remote_=<optimized out>) at ./build_docker/./src/Common/Exception.cpp:105
2024-07-01 09:03:19 No locals.
2024-07-01 09:03:19 #7  0x0000562e70c3d740 in DB::Exception::Exception(PreformattedMessage&&, int) ()
2024-07-01 09:03:20 No symbol table info available.
2024-07-01 09:03:20 #8  0x0000562e70c3d0e0 in DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(int, FormatStringHelperImpl<std::__1::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&) ()
2024-07-01 09:03:20 No symbol table info available.
2024-07-01 09:03:20 #9  0x0000562e8291ba95 in DB::MergeTask::VerticalMergeStage::prepareVerticalMergeForAllColumns (this=this@entry=0x722000b32d18) at ./build_docker/./src/Storages/MergeTree/MergeTask.cpp:562
2024-07-01 09:03:20         reread_buf_ref = <optimized out>
2024-07-01 09:03:20         sum_input_rows_exact = 2
2024-07-01 09:03:20         input_rows_filtered = 2
2024-07-01 09:03:20         rows_sources_count = 0
2024-07-01 09:03:20         wbuf_readable = <optimized out>
2024-07-01 09:03:20         reread_buf = <optimized out>
2024-07-01 09:03:20         reread_buffer_raw = <optimized out>
2024-07-01 09:03:20         reread_buffer_from_file = <optimized out>
2024-07-01 09:03:20         settings = <optimized out>
2024-07-01 09:03:20         max_delayed_streams = <optimized out>
2024-07-01 09:03:20         all_parts_on_remote_disks = <optimized out>
2024-07-01 09:03:20 #10 0x0000562e8292215c in DB::MergeTask::VerticalMergeStage::execute (this=<optimized out>) at ./build_docker/./src/Storages/MergeTree/MergeTask.cpp:891
2024-07-01 09:03:21 No locals.
2024-07-01 09:03:21 #11 0x0000562e82921b69 in DB::MergeTask::execute (this=0x721800ad7978) at ./build_docker/./src/Storages/MergeTree/MergeTask.cpp:951
2024-07-01 09:03:21         next_stage_context = {__ptr_ = 0x562e70c0af60 <__tsan_atomic64_fetch_add+112>, __cntrl_ = 0x4}
2024-07-01 09:03:21 #12 0x0000562e82e3f469 in DB::MergePlainMergeTreeTask::executeStep (this=this@entry=0x724c012cf918) at ./build_docker/./src/Storages/MergeTree/MergePlainMergeTreeTask.cpp:57
2024-07-01 09:03:21         profile_events_scope = {<boost::noncopyable_::noncopyable> = {<boost::noncopyable_::base_token> = {<No data fields>}, <No data fields>}, performance_counters_holder = {__ptr_ = {<std::__1::__compressed_pair_elem<ProfileEvents::Counters*, 0, false>> = {__value_ = 0x0}, <std::__1::__compressed_pair_elem<std::__1::default_delete<ProfileEvents::Counters>, 1, true>> = {<std::__1::default_delete<ProfileEvents::Counters>> = {<No data fields>}, <No data fields>}, <No data fields>}}, performance_counters_scope = 0x724c012cfa60, previous_counters_scope = 0x7f3cbe371698}
2024-07-01 09:03:21         switcher = {<std::__1::__optional_move_assign_base<DB::ThreadGroupSwitcher, false>> = {<std::__1::__optional_copy_assign_base<DB::ThreadGroupSwitcher, false>> = {<std::__1::__optional_move_base<DB::ThreadGroupSwitcher, false>> = {<std::__1::__optional_copy_base<DB::ThreadGroupSwitcher, false>> = {<std::__1::__optional_storage_base<DB::ThreadGroupSwitcher, false>> = {<std::__1::__optional_destruct_base<DB::ThreadGroupSwitcher, false>> = {{__null_state_ = 24 '\030', __val_ = {<boost::noncopyable_::noncopyable> = {<boost::noncopyable_::base_token> = {<No data fields>}, <No data fields>}, prev_thread_group = {__ptr_ = 0x725002f42618, __cntrl_ = 0x725002f42600}}}, __engaged_ = true}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, <std::__1::__sfinae_ctor_base<false, false>> = {<No data fields>}, <std::__1::__sfinae_assign_base<false, false>> = {<No data fields>}, <No data fields>}
2024-07-01 09:03:21 #13 0x0000562e82e072a6 in DB::executeHere (task=...) at ./src/Storages/MergeTree/MergePlainMergeTreeTask.h:104
2024-07-01 09:03:21 No locals.
2024-07-01 09:03:21 #14 DB::StorageMergeTree::merge (this=this@entry=0x727c004fae40, aggressive=<optimized out>, partition_id=..., final=true, deduplicate=false, deduplicate_by_columns=..., cleanup=false, txn=..., out_disable_reason=..., optimize_skip_merged_partitions=<optimized out>) at ./build_docker/./src/Storages/StorageMergeTree.cpp:1170
2024-07-01 09:03:23         table_lock_holder = {__ptr_ = 0x721801220b98, __cntrl_ = 0x721801220b80}
2024-07-01 09:03:23         select_decision = DB::SelectPartsDecision::SELECTED
2024-07-01 09:03:23         f = {<std::__1::__function::__maybe_derive_from_unary_function<void (bool)>> = {<std::__1::__unary_function_keep_layout_base<bool, void>> = {<No data fields>}, <No data fields>}, <std::__1::__function::__maybe_derive_from_binary_function<void (bool)>> = {<No data fields>}, __f_ = {__buf_ = {__small = "X\272O\000|r\000\000\001\376\214\000\br\000", __large = 0x727c004fba58}, __invoker_ = {__call_ = 0x562e82e28be0 <std::__1::__function::__policy_invoker<void (bool)>::__call_impl<std::__1::__function::__default_alloc_func<DB::StorageMergeTree::merge(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, bool, std::__1::shared_ptr<DB::MergeTreeTransaction> const&, PreformattedMessage&, bool)::$_0, void (bool)> >(std::__1::__function::__policy_storage const*, bool)>}, __policy_ = 0x562e8a7a1930 <std::__1::__function::__policy::__choose_policy[abi:v15000]<std::__1::__function::__default_alloc_func<DB::StorageMergeTree::merge(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, bool, std::__1::shared_ptr<DB::MergeTreeTransaction> const&, PreformattedMessage&, bool)::$_0, void (bool)> >(std::__1::integral_constant<bool, true>)::__policy_>}}
2024-07-01 09:03:23         task = {__ptr_ = 0x724c012cf918, __cntrl_ = 0x724c012cf900}
2024-07-01 09:03:23         metadata_snapshot = {__ptr_ = 0x727400f83400, __cntrl_ = 0x7208006d64c0}
2024-07-01 09:03:23         merge_mutate_entry = {__ptr_ = 0x7214003ea408, __cntrl_ = 0x7214003ea3f0}
2024-07-01 09:03:23 #15 0x0000562e82e0ddba in DB::StorageMergeTree::optimize (this=0x727c004fae40, partition=..., final=<optimized out>, deduplicate=false, deduplicate_by_columns=..., cleanup=<optimized out>, local_context=...) at ./build_docker/./src/Storages/StorageMergeTree.cpp:1582
2024-07-01 09:03:23         partition_id = @0x720c00a76b00: {static __endian_factor = 1, __r_ = {<std::__1::__compressed_pair_elem<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__rep, 0, false>> = {__value_ = {{__l = {__data_ = 0x31 <error: Cannot access memory at address 0x31>, __size_ = 0, __cap_ = 72057594037927936, __is_long_ = 0}, __s = {__data_ = "1", '\000' <repeats 21 times>, __padding_ = 0x720c00a76b17 "\001ql", __size_ = 1 '\001', __is_long_ = 0 '\000'}, __r = {__words = {49, 0, 72057594037927936}}}}}, <std::__1::__compressed_pair_elem<std::__1::allocator<char>, 1, true>> = {<std::__1::allocator<char>> = {<std::__1::__non_trivial_if<true, std::__1::allocator<char> >> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, static npos = 18446744073709551615}
2024-07-01 09:03:23         __range2 = @0x7f3cbe36b7d8: {__table_ = {__bucket_list_ = {__ptr_ = {<std::__1::__compressed_pair_elem<std::__1::__hash_node_base<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>*>**, 0, false>> = {__value_ = 0x720c01f1e7a0}, <std::__1::__compressed_pair_elem<std::__1::__bucket_list_deallocator<std::__1::allocator<std::__1::__hash_node_base<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>*>*> >, 1, false>> = {__value_ = {__data_ = {<std::__1::__compressed_pair_elem<unsigned long, 0, false>> = {__value_ = 5}, <std::__1::__compressed_pair_elem<std::__1::allocator<std::__1::__hash_node_base<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>*>*>, 1, true>> = {<std::__1::allocator<std::__1::__hash_node_base<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>*>*>> = {<std::__1::__non_trivial_if<true, std::__1::allocator<std::__1::__hash_node_base<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>*>*> >> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}}}, <No data fields>}}, __p1_ = {<std::__1::__compressed_pair_elem<std::__1::__hash_node_base<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>*>, 0, false>> = {__value_ = {__next_ = 0x720c01e24f30}}, <std::__1::__compressed_pair_elem<std::__1::allocator<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*> >, 1, true>> = {<std::__1::allocator<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*> >> = {<std::__1::__non_trivial_if<true, std::__1::allocator<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*> > >> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, __p2_ = {<std::__1::__compressed_pair_elem<unsigned long, 0, false>> = {__value_ = 5}, <std::__1::__compressed_pair_elem<std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, 1, true>> = {<std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >> = {<std::__1::__unary_function_keep_layout_base<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, unsigned long>> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}, __p3_ = {<std::__1::__compressed_pair_elem<float, 0, false>> = {__value_ = 1}, <std::__1::__compressed_pair_elem<std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, 1, true>> = {<std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >> = {<std::__1::__binary_function_keep_layout_base<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool>> = {<No data fields>}, <No data fields>}, <No data fields>}, <No data fields>}}}
```

**Review by vdimir — APPROVED** [review]

**Comment by pkit** [comment]
Was the bug introduced in https://github.com/ClickHouse/ClickHouse/pull/59163 or earlier?

**Comment by canhld94** [comment]
Yes, it’s introduced in #59163

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by pkit** [comment]
Okay, this will lead to ClickHouse DDoS, if the table is not deleted, while ingestion happens: ClickHouse will try to re-create the .sst files, fail to do it, create file again, again and again. And then the whole database will die with "too many open files".

**Comment by canhld94** [comment]
@pkit can you demonstrate with an example? IIUC if sink can't create sst file it will throw exception instead of retrying.

**Comment by pkit** [comment]
> @pkit can you demonstrate with an example? IIUC if sink can't create sst file it will throw exception instead of retrying.

I will try to get it reproduced in an isolated env, but the gist of it was: upgrade from 24.2.1 to 24.5.1 with existing ttl-ed RocksDB.
RocksDB tables were fed from MVs.
And the end result was: 100k open sst files per table.
System worked well on 24.2.1 for months before.

**Comment by canhld94** [comment]
@pufit in this case, I suspect the issue is you're inserting to rocksdb table in many small blocks.
FYI, you can set the table to use normal sink with memtable by changing table setting "optimize_for_bulk_load" to 0.

**Comment by pkit** [comment]
> you're inserting to rocksdb table in many small blocks

Yes, obviously, we use it as k/v store. Which means fast insert and fast select of a single key.
Imagine you have a stream of events, and you want to join it with another stream.
Keeping some number of events (TTL-ed) as RocksDB enables that use case.
I see you've changed the default which lead to a total collapse too... Eh... :)

**Pull request merged**